orc8r/cloud/go/services/state/indexer/reindex/reindexer_singleton_test.go

```diff
@@ -0,0 +1,169 @@
+/*
+Copyright 2020 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reindex_test
+
+import (
+	"context"
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+
+	"magma/orc8r/cloud/go/clock"
+	configurator_test_init "magma/orc8r/cloud/go/services/configurator/test_init"
+	configurator_test "magma/orc8r/cloud/go/services/configurator/test_utils"
+	device_test_init "magma/orc8r/cloud/go/services/device/test_init"
+	directoryd_types "magma/orc8r/cloud/go/services/directoryd/types"
+	"magma/orc8r/cloud/go/services/orchestrator/obsidian/models"
+	"magma/orc8r/cloud/go/services/state/indexer"
+	"magma/orc8r/cloud/go/services/state/indexer/reindex"
+	state_test_init "magma/orc8r/cloud/go/services/state/test_init"
+	state_test "magma/orc8r/cloud/go/services/state/test_utils"
+)
+
+func TestSingletonRun(t *testing.T) {
+	// Make nullimpotent calls to handle code coverage indeterminacy
+	reindex.TestHookReindexSuccess()
+	reindex.TestHookReindexDone()
+
+	// Writes to channel after completing a job
+	reindexSuccessNum, reindexDoneNum := 0, 0
+	ch := make(chan interface{})
+
+	reindex.TestHookReindexSuccess = func() {
+		reindexSuccessNum += 1
+	}
+	defer func() { reindex.TestHookReindexSuccess = func() {} }()
+
+	reindex.TestHookReindexDone = func() {
+		reindexDoneNum += 1
+		ch <- nil
+	}
+	defer func() { reindex.TestHookReindexSuccess = func() {} }()
+
+	clock.SkipSleeps(t)
+	defer clock.ResumeSleeps(t)
+
+	r := initSingletonReindexTest(t)
+	ctx, cancel := context.WithCancel(context.Background())
+	go r.Run(ctx)
+	defer cancel()
+
+	// Single indexer
+	idx0 := getIndexer(id0, zero, version0, true)
+	idx0.On("GetTypes").Return(allTypes).Once()
+	// Register indexers
+	register(t, idx0)
+
+	// Check
+	recvCh(t, ch)
+	// QUESTION: is this not recommended? We check that nothing else is sent for a whole second.
```

Comment:
Hm. It's not the end of the world since tests get parallelized. But I think in this case it's probably not necessary -- just let the subsequent recvCh handle the expectation

Reply:
Oh since we are resetting the index between each test piece, the next emitted `ch <- nil` from the past test might've normally run and not have been tested. This is something that sometimes happened while testing bad code. 

Kinda confusing, I can explain better if that's not sufficient/any thoughts?

Reply:
Not following, can you explain more here or in 1-1

Reply:
## Example
For the same test case response where `reindexer_singleton` sends `ch -> nil` twice (where it should only send once), stack of code being run could be:
### In case where we have `recvNoCh(t, ch)`
1) `register(t, idx0)`
2) `ch-> nil` (received by `recvCh(t, ch)`)
3) `ch -> nil` (received by `recvNoCh(t, ch)`) ==> this causes error, as expected

### In case where we don't have `recvNoCh(t, ch)`
1) `register(t, idx0)`
2) `ch-> nil` (received by `recvCh(t, ch)`)
...
3)  `register(t, idx0a)`

Results in no error since the second `ch-> nil` would not have time to be sent before `register(t, idx0a)` happened

Reply:
Ah I see. Okay that makes sense, let's keep it then

P.S. feel free to re-request a review when ready